DOC-9205 - KV Operations dev-reorg (#2275)
couchbase/docs-server: modules/guides/pages/reading-data.adoc

= Reading Data
:description: How to read documents in Couchbase.
:page-pagination: full
:page-topic-type: guide
:page-toclevels: 2
:tabs:
:github: Click the GitHub button icon:github[] to view this code in context.

include::partial$example-attributes.adoc[]

[abstract]
{description}

> 

Reviewer: For consistency with my sections

```suggestion

== Introduction

```

Retrieving documents by ID is the fastest and simplest way to read xref:learn:data/data.adoc[data] in Couchbase.
The xref:learn:services-and-indexes/services/data-service.adoc[Key Value (KV) or Data Service] allows you to retrieve a full document when you need to fetch all of the data stored. 
However, in instances where this can be costly and unnecessary, Couchbase also provides access to specific _paths_ within a document.

> 

Reviewer: Include links to the client tools and warning about altering your data

```suggestion

include::partial$clients.adoc[tag=refs]

include::partial$data-warning.adoc[]

```

== Reading a Document

> To read a single document in Couchbase perform a `get` operation.

Reviewer: This makes `get` look like a command or option -- remove markup

```suggestion
To read a single document in Couchbase, perform a get operation.
```

[{tabs}]
====
cbc::
+ 
--
Use the `cbc cat` command to retrieve a document by ID and output its data.

{kv-get-example}

[source,shell]
----
include::example$kv/kv-cbc.sh[tag=cbc-kv-get]
----

.Result
[source,console]
----
hotel-123            CAS=0x16ba896b78930000, Flags=0x0, Size=567, Datatype=0x01(JSON)

{
  "id": 123,
  "name": "Medway Youth Hostel",
  "address": "Capstone Road, ME7 3JE",
  "url": "http://www.yha.org.uk",
  "geo": {
    "lat": 51.35785,
    "lon": 0.55818,
    "accuracy": "RANGE_INTERPOLATED"
  },
  "country": "United Kingdom",
  "city": "Medway",
  "state": null,
  "reviews": [
    {
      "content": "This was our 2nd trip here and we enjoyed it more than last year.",
      "author": "Ozella Sipes",
      "date": "2021-11-17T17:35:05.351Z"
    }
  ],
  "vacancy": true,
  "description": "40 bed summer hostel about 3 miles from Gillingham."
}
----

The output has been prettified for readability.

NOTE: If the document cannot be found, [.cmd]`cbc` will return a `LCB_ERR_DOCUMENT_NOT_FOUND` error. 

> --

Reviewer: Add link to relevant reference doc.

.NET::
+ 
--
. Use the `GetAsync()` method to retrieve a document by ID from the `travel-sample.inventory.hotel` keyspace.

> . A `GetResult` object is returned, which includes the `content`, `cas` value, and other valuable metadata.

Reviewer: Each numbered step in a procedure should contain a "do". The second step here actually just describes the results of the first step. There's a couple of ways we could fix this, but I think the easiest is to turn the second step into an unnumbered paragraph following on from the previous step. However, that means the first step becomes the only numbered step in the procedure. This looks strange, and I'm not quite sure how to handle it at the moment. For now I suggest just removing all the numbering from this short procedure.

```suggestion
Use the `GetAsync()` method to retrieve a document by ID from the `travel-sample.inventory.hotel` keyspace.

A `GetResult` object is returned, which includes the `content`, `cas` value, and other valuable metadata.
```

{kv-get-example}

[source,csharp]
----
include::dotnet-sdk:hello-world:example$KvHelloWorldScoped.csx[tag=kv-get,indent=0]
----

NOTE: If the document doesn't exist, the SDK will return a `DocumentNotFoundException` error. 

{github}

> --

Reviewer: Add link to relevant reference doc.

Java::
+ 
--
. Use the `get()` method to retrieve a document by ID from the `travel-sample.inventory.hotel` keyspace.

> . A `GetResult` object is returned, which includes the `content`, `cas` value, and other valuable metadata.

Reviewer: ```suggestion
Use the `get()` method to retrieve a document by ID from the `travel-sample.inventory.hotel` keyspace.

A `GetResult` object is returned, which includes the `content`, `cas` value, and other valuable metadata.
```

{kv-get-example}

[source,java]
----
include::java-sdk:hello-world:example$KvHelloWorldScoped.java[tag=kv-get,indent=0]
----

NOTE: If the document doesn't exist, the SDK will return a `DocumentNotFoundException` error. 

{github}

> --

Reviewer: Add link to relevant reference doc.

Node.js::
+ 
--
. Use the `get()` function to retrieve a document by ID from the `travel-sample.inventory.hotel` keyspace.

> . A `GetResult` promise is returned, which includes the `content`, `cas` value, and other valuable metadata.

Reviewer: ```suggestion
Use the `get()` function to retrieve a document by ID from the `travel-sample.inventory.hotel` keyspace.

A `GetResult` promise is returned, which includes the `content`, `cas` value, and other valuable metadata.
```

{kv-get-example}

[source,nodejs]
----
include::nodejs-sdk:hello-world:example$kv-hello-world-scoped.js[tag=kv-get,indent=0]
----

NOTE: If the document doesn't exist, the SDK will return a `DocumentNotFoundError` error. 

{github}

> --

Reviewer: Add link to relevant reference doc.

Python::
+ 
--
. Use the `get()` function to retrieve a document by ID from the `travel-sample.inventory.hotel` keyspace.

> . A `GetResult` object is returned, which includes the `content`, `cas` value, and other valuable metadata.

Reviewer: ```suggestion
Use the `get()` function to retrieve a document by ID from the `travel-sample.inventory.hotel` keyspace.

A `GetResult` object is returned, which includes the `content`, `cas` value, and other valuable metadata.
```

{kv-get-example}

[source,python]
----
include::python-sdk:hello-world:example$kv_hello_world_scoped.py[tag=kv-get,indent=0]
----

NOTE: If the document doesn't exist, the SDK will return a `DocumentNotFoundException` error.

{github}

> --

Reviewer: Add link to relevant reference doc.

====

=== Reading with Options

> It is also possible to add `options` to a get operation.

Reviewer: Start procedures with "To:"

```suggestion
To specify further parameters, such as expiry, add options to the get operation.
```

[{tabs}]
====
cbc::
+ 
--

> Use the `cbc cat` command to retrieve a document by ID and pass some options.

Reviewer: ```suggestion
Use the `cbc cat` command to retrieve a document by ID and pass options as required.
```

The example below uses an `--expiry` option for the `cat` command which adds an expiry of 60 seconds to the `hotel-123` document.

[source,shell]
----
include::example$kv/kv-cbc.sh[tag=cbc-kv-get-with-opts]
----

.Result
[source,console]
----
hotel-123            CAS=0x16bcec98e00c0000, Flags=0x0, Size=567, Datatype=0x01(JSON)

{
  "id": 123,
  "name": "Medway Youth Hostel",
  "address": "Capstone Road, ME7 3JE",
  "url": "http://www.yha.org.uk",
  "geo": {
    "lat": 51.35785,
    "lon": 0.55818,
    "accuracy": "RANGE_INTERPOLATED"
  },
  "country": "United Kingdom",
  "city": "Medway",
  "state": null,
  "reviews": [
    {
      "content": "This was our 2nd trip here and we enjoyed it more than last year.",
      "author": "Ozella Sipes",
      "date": "2021-11-17T17:35:05.351Z"
    }
  ],
  "vacancy": true,
  "description": "40 bed summer hostel about 3 miles from Gillingham."
}
----

The output has been prettified for readability.

> --

Reviewer: Add link to relevant reference doc.

.NET::
+ 
--
. Pass an `Expiry()` option to the `GetAsync()` method when retrieving a document. 

> . A `GetResult` object is returned, which includes the `content`, `cas` value, and `expiry` metadata.

Reviewer: Generalize instructions, and remove numbering as there's only one step.

```suggestion
Pass any required options to the `GetAsync()` method when retrieving a document. 

A `GetResult` object is returned, which may include extra metadata, depending on the options passed.
```

{kv-get-example-with-options}

[source,csharp]
----
include::dotnet-sdk:hello-world:example$KvHelloWorldScoped.csx[tag=kv-get-with-opts,indent=0]
----

{github}

> --

Reviewer: Add link to relevant reference doc.

Java::
+ 
--
. Pass a `withExpiry(true)` option to the `GetAsync()` method when retrieving a document.

> . A `GetResult` object is returned, which includes the `content`, `cas` value, and `expiry` metadata.

Reviewer: Generalize instructions, and remove numbering as there's only one step.

```suggestion
Pass any required options to the `GetAsync()` method when retrieving a document.

A `GetResult` object is returned, which may include extra metadata, depending on the options passed.
```

{kv-get-example-with-options}

[source,java]
----
include::java-sdk:hello-world:example$KvHelloWorldScoped.java[tag=kv-get-with-opts,indent=0]
----

{github}

> --

Reviewer: Add link to relevant reference doc.

Node.js::
+ 
--
. Pass a `withExpiry: true` option to the `GetAsync()` method when retrieving a document.

> . A `GetResult` promise is returned, which includes the `content`, `cas` value, and `expiry` metadata.

Reviewer: Generalize instructions, and remove numbering as there's only one step.

```suggestion
Pass any required options to the `GetAsync()` method when retrieving a document.

A `GetResult` promise is returned, which may include extra metadata, depending on the options passed.
```

{kv-get-example-with-options}

[source,nodejs]
----
include::nodejs-sdk:hello-world:example$kv-hello-world-scoped.js[tag=kv-get-with-opts,indent=0]
----

{github}

> --

Reviewer: Add link to relevant reference doc.

Python::
+ 
--
. Pass a `with_expiry=True` option to the `GetAsync()` method when retrieving a document.

> . A `GetResult` object is returned, which includes the `content`, `cas` value, and `expiry` metadata.

Reviewer: Generalize instructions, and remove numbering as there's only one step.

```suggestion
Pass any required options to the `GetAsync()` method when retrieving a document.

A `GetResult` object is returned, which may include extra metadata, depending on the options passed.
```

{kv-get-example-with-options}

[source,python]
----
include::python-sdk:hello-world:example$kv_hello_world_scoped.py[tag=kv-get-with-opts,indent=0]
----

{github}

> --

Reviewer: Add link to relevant reference doc.

====

== Reading a Sub-Document

JSON documents can contain a lot of nested data -- which might not necessarily need to be accessed all at once. 
Reading full documents to access a field or two is not ideal and could cause performance issues in your application. 
Instead, a better practice would be to access specific paths, or _sub-documents_, to perform more efficient read operations.

> To read specific fields or document paths you can perform a sub-document `get` operation.

Reviewer: "Get" in this case is not a command or option

```suggestion
To fetch a specific field inside a document, you can perform a sub-document get operation.
```

[{tabs}]

> ====

Reviewer: Add links to relevant reference docs for each tab.

cbc-subdoc::
+ 
--
. Connect to the `cbc-subdoc` interactive shell.

. Use the `get` command to access specific fields in a JSON document with the `--path` argument. 

The example below fetches the `geo` data from the `airport_1254` document.

.airport_1254
[source,json]
----
{
  "id": 1254,
//  ...
  "geo": { 
    "lat": 50.962097,
    "lon": 1.954764,
    "alt": 12
  }
}
----

[source,console]
----
cbc-subdoc -u Administrator -P password -U couchbase://localhost/travel-sample
subdoc> get airport_1254 --path geo
----

.Result
[source,console]
----
airport_1254         CAS=0x16b815068df80000
0. Size=43, RC=LCB_SUCCESS (0)
{"lat":50.962097,"lon":1.954764,"alt":12.0}
----

NOTE: If the path cannot be found, `cbc-subdoc` will return a `LCB_ERR_SUBDOC_PATH_NOT_FOUND` error. 
--

.NET::
+ 
--
. To fetch a specific field inside a document, call the `LookupInAsync()` method, which takes a document ID and an IEnumerable containing `LookUpInSpec` objects. 
`LookUpInSpec` represents the sub-operation to be performed within the lookup.

> . A `LookupInResult` object is returned containing the result and metadata relevant to the sub-document get operation.

Reviewer: Each numbered step should be a "do".

```suggestion
. Call the `LookupInAsync()` method, which takes a document ID and an IEnumerable containing `LookUpInSpec` objects.

. Use the `LookUpInSpec` object to specify the sub-operation to be performed within the lookup.

A `LookupInResult` object is returned, containing the result and metadata relevant to the sub-document get operation.
```

{kv-subdoc-get-example}

[source,csharp]
----
include::dotnet-sdk:hello-world:example$KvHelloWorldScoped.csx[tag=kv-get-subdoc,indent=0]
----

{github}
--

Java::
+ 
--
. To fetch a specific field inside a document, call the `lookupIn()` method, which takes a document ID and an array of `LookUpInSpec` objects. 
`LookUpInSpec` represents the sub-operation to be performed within the lookup.

> . A `LookupInResult` object is returned containing the result and metadata relevant to the sub-document get operation.

Reviewer: Each numbered step should be a "do".

```suggestion
. Call the `lookupIn()` method, which takes a document ID and an array of `LookUpInSpec` objects.

. Use the `LookUpInSpec` object to represent the sub-operation to be performed within the lookup.

A `LookupInResult` object is returned containing the result and metadata relevant to the sub-document get operation.
```

{kv-subdoc-get-example}

[source,java]
----
include::java-sdk:hello-world:example$KvHelloWorldScoped.java[tag=kv-get-subdoc,indent=0]
----

NOTE: If the document path can't be found, the SDK will return a `PathNotFoundException` error.

{github}
--

Node.js::
+ 
--
. To fetch a specific field inside a document, call the `lookupIn()` function, which takes a document ID and an array of `LookUpInSpec` objects. 
`LookUpInSpec` represents the sub-operation to be performed within the lookup.

> . A `LookupInResult` promise is returned containing the result and metadata relevant to the sub-document get operation.

Reviewer: Each numbered step should be a "do".

```suggestion
. Call the `lookupIn()` function, which takes a document ID and an array of `LookUpInSpec` objects.

. Use the `LookUpInSpec` object to specify the sub-operation to be performed within the lookup.

A `LookupInResult` promise is returned containing the result and metadata relevant to the sub-document get operation.
```

{kv-subdoc-get-example}

[source,nodejs,indent=0]
----
include::nodejs-sdk:hello-world:example$kv-hello-world-scoped.js[tag=kv-get-subdoc]
----

{github}
--

Python::
+ 
--
. To fetch a specific field inside a document, call the `lookup_in()` function, which takes a document ID and a list of `LookUpInSpec` objects. 
`LookUpInSpec` represents the sub-operation to be performed within the lookup.

> . A `LookupInResult` object is returned containing the result and metadata relevant to the sub-document get operation.

Reviewer: Each numbered step should be a "do".

```suggestion
. Call the `lookup_in()` function, which takes a document ID and a list of `LookUpInSpec` objects.

. Use the `LookUpInSpec` object to represent the sub-operation to be performed within the lookup.

A `LookupInResult` object is returned containing the result and metadata relevant to the sub-document get operation.
```
